internal/cmd/tls.go

```diff
@@ -0,0 +1,255 @@
+		tlsSelfSign(),
+		tlsCustom(),
+		tlsACME(),
+		tlsDisable(),
+	)
+	return cmd
+}
+
+func tlsSelfSign() *cobra.Command {
+	var (
+		hosts []string
+	)
+	cmd := &cobra.Command{
+		Use:     "self-sign",
+		Short:   "Generate self-signed certificate for Coder",
+		Long:    "Generate self-signed certificates for Coder. Self-signed certificates are automatically renewed by Coder",
+		Example: "tls self-sign --hosts a.example.com --hosts b.example.com --hosts c.example.com",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+			client, err := newClient(ctx, true)
+			if err != nil {
+				return err
+			}
+
+			if len(hosts) == 0 {
+				return ErrMissingRequiredParameter{"hosts"}
+			}
+
+			resp, err := client.GenerateSelfSignedCertificate(ctx, coder.SelfSignedCertificateRequest{
+				Hosts: hosts,
+			})
+			if err != nil {
+				return err
+			}
+
+			if !resp.Success {
+				return xerrors.New("Failed to generate self-signed certificate")
+			}
+
+			clog.LogSuccess("Generated self-signed certificate")
+			return nil
+		},
+	}
+
+	cmd.Flags().StringArrayVar(&hosts, "hosts", []string{}, "Hostnames and/or IPs to generate self-signed certificate for")
+
+	return cmd
+}
+
+func tlsCustom() *cobra.Command {
+	var (
+		certFilepath string
+		keyFilepath  string
+	)
+
+	cmd := &cobra.Command{
+		Use:     "custom",
+		Short:   "Upload custom PEM encoded certificate and key files",
+		Example: "tls custom --cert /path/to/cert.pem --key /path/to/key.pem",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+			client, err := newClient(ctx, true)
+			if err != nil {
+				return err
+			}
+
+			cert, err := os.ReadFile(certFilepath)
+			if err != nil {
+				return err
+			}
+
+			key, err := os.ReadFile(keyFilepath)
+			if err != nil {
+				return err
+			}
+
+			req := coder.UploadCertificateRequest{
+				Certificate: cert,
+				PrivateKey:  key,
+			}
+
+			resp, err := client.UploadCustomCertificate(ctx, req)
+			if err != nil {
+				return err
+			}
+
+			if !resp.Success {
+				return xerrors.New("Failed to apply custom TLS certificate")
+			}
+
+			clog.LogSuccess("Uploaded custom TLS certificate")
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVarP(&certFilepath, "cert", "c", "", "Full path to a PEM encoded certificate file")
+	cmd.Flags().StringVarP(&keyFilepath, "key", "k", "", "Full path to a PEM encoded private key file")
+	return cmd
+}
+
+func tlsACME() *cobra.Command {
+	var (
+		showProviderInfo bool
+
+		agreeTOS    bool
+		email       string
+		domains     []string
+		dnsProvider string
+		credentials map[string]string
+	)
+	cmd := &cobra.Command{
+		Use:   "acme",
+		Short: "Generate certificate via Let's Encrypt",
+		Example: `
+tls acme --info
+tls acme --email me@example.com --domains a.example.com --domains b.example.com --provider route53 --credentials AWS_ACCESS_KEY_ID=your-key-id --credentials AWS_SECRET_ACCESS_KEY=your-secret-key --credentials AWS_REGION=your-region`,
```

Comment:
it would be nice if we could pass credentials from environment variables if possible, for example: `tls acme --credentials AWS_ACCESS_KEY_ID` would read the `AWS_ACCESS_KEY_ID` environment variable and use that as a credential (or event do that automatically if those variables are set). This is because AWS users often have these environment variables set, and they are respected by tools they're using.

I think if you set the env vars, then the `aws cli` command will use them: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html - and it would be nice if we could do that with this tool too, like if someone specifies route53, then use their default env vars if set.

Feel free to do that in a separate PR (or not at all) though! Not a big deal, just a UX enhancement